microsoft-office-365-feed/operations.py

```diff
@@ -0,0 +1,227 @@
+                          captured_value['ClientRequestId'][0]
+                    response = requests.get(
+                        url=url,
+                        verify=params.get('verify_ssl')
+                    )
+                if response.status_code == 200:
+                    data = response.json()
+                    if "latest" in data:
+                        result.append(data)
+                    else:
+                        indicators = [i for i in data if 'ips' in i or 'urls' in i]  # filter empty entries
+                        result.extend(indicators)
+                else:
+                    response.raise_for_status()
+            except requests.exceptions.SSLError:
+                logger.error('An SSL error occurred.')
+                raise ConnectorError('An SSL error occurred.')
+            except requests.exceptions.ConnectionError:
+                logger.error('A connection error occurred.')
+                raise ConnectorError('A connection error occurred.')
+            except Exception as err:
+                logger.error(err)
+                raise ConnectorError(err)
+        return result
+    except Exception as Err:
+        logger.error(str(Err))
+        raise ConnectorError(str(Err))
+
+
+def get_latest_version(config):
+    try:
+        client_request_id = str(uuid.uuid4())
+        region = config.get('regions')
+        params = {"urls_list": build_urls_dict(region, services_list=["All"], unique_id=client_request_id,
+                                               endpoint="version"),
+                  "verify_ssl": config.get("verify_ssl")}
+
+        resp = make_api_call(params)
+        return resp
+    except Exception as Err:
+        logger.error(str(Err))
+        raise ConnectorError(str(Err))
+
+
+def fetch_last_updated_time(config, last_pull_time):
+    region = config.get('regions')
+    latest_version_list = get_latest_version(config)
+    for version in latest_version_list:
+        if version["instance"] == region:
+            latest_time = version["latest"]
+            try:
+                old_version_ts = datetime.datetime.strptime(str(last_pull_time), "%Y-%m-%dT%H:%M:%SZ").timestamp()
+            except:
+                old_version_ts = last_pull_time
+            latest_version_ts = int(datetime.datetime.strptime(latest_time, "%Y%m%d%H").timestamp())
+            if old_version_ts < latest_version_ts:
+                return True, latest_version_list
+    return False, None
+
+
+def build_response(indicator_fields, indicator_type_lower, limit=-1):
+    iterator = make_api_call(indicator_fields)
+    result = indicator_fields["urls_list"][0]
+    # filter indicator_type specific records
+    if not indicator_type_lower == 'both':
+        iterator = [i for i in iterator if indicator_type_lower in i]
+    indicators = []
+
+    for item in iterator:
+        if indicator_type_lower == 'both':
+            records = {'IP': item.get('ips', []),
```

Comment:
why do we need this grouping? can we not return the api response as is? iterating over the complete list again in expensive and can be avoided unless necessary